### PR TITLE
restrict columns to read for pandas.read_parquet

### LINKED ISSUE #18154: Enable to restrict columns for pandas.read_parquet
#### Problem description

In pandas 0.21 the top level funtion read_parquet() was introduced. Both available engines fastparquet and pyarrow support the specifications of columns to read. If you are only interested in certain columns of a dataframe this reduces the io.

- Fastparquet https://github.com/dask/fastparquet/blob/master/docs/source/quickstart.rst#reading

- PyArrow http://pyarrow-xhochy.readthedocs.io/en/latest/pyarrow.parquet.html#pyarrow.parquet.read_table

It should be also possible to specify the columns in pandas.read_parquet().

### PR DESCRIPTION
**Pull request by hoffmann**
- [ x ] closes #18154
- [ x ] tests added / passed
- [ x ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/tests/io/test_parquet.py [hunk 1]
```diff
@@ -282,6 +282,17 @@ def test_compression(self, engine, compression):
         df = pd.DataFrame({'A': [1, 2, 3]})
         self.check_round_trip(df, engine, compression=compression)
 
+    def test_read_columns(self, engine, fp):
+        df = pd.DataFrame({'string': list('abc'),
```
Comment (gfyoung) [inline pandas/tests/io/test_parquet.py]: Reference issue number above.

File: pandas/io/parquet.py [hunk 2]
```diff
@@ -188,6 +188,8 @@ def read_parquet(path, engine='auto', **kwargs):
     ----------
     path : string
         File path
+    columns: list
```
Comment (gfyoung) [inline pandas/io/parquet.py]: Write out what the default is too i.e. "list, default None"

**Comment by gfyoung** [comment]
Will need a `whatsnew` note in `0.22.0`

File: doc/source/whatsnew/v0.22.0.txt [hunk 3]
```diff
@@ -109,7 +109,7 @@ I/O
 ^^^
 
 - :func:`read_html` now rewinds seekable IO objects after parse failure, before attempting to parse with a new parser. If a parser errors and the object is non-seekable, an informative error is raised suggesting the use of a different parser (:issue:`17975`)
--
+- :func:`read_parquet` now allows to specify the columns to read from a parquet file (:issue:`18154`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.22.0.txt]: you can put this on 0.21.1

File: pandas/io/parquet.py [hunk 4]
```diff
@@ -188,6 +188,8 @@ def read_parquet(path, engine='auto', **kwargs):
     ----------
     path : string
         File path
+    columns: list, default=None
```
Comment (jreback) [inline pandas/io/parquet.py]: add a version added tag

File: doc/source/whatsnew/v0.22.0.txt [hunk 5]
```diff
@@ -109,7 +109,7 @@ I/O
 ^^^
 
 - :func:`read_html` now rewinds seekable IO objects after parse failure, before attempting to parse with a new parser. If a parser errors and the object is non-seekable, an informative error is raised suggesting the use of a different parser (:issue:`17975`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.22.0.txt]: can you add a small example in the docs in io.rst as well.

**Review by jreback — CHANGES_REQUESTED** [review]
doc comments, lgtm. otherwise.

File: pandas/io/parquet.py (hunk 4 above)
Comment (hoffmann) [reply pandas/io/parquet.py]: done

File: doc/source/whatsnew/v0.22.0.txt (hunk 5 above)
Comment (hoffmann) [reply doc/source/whatsnew/v0.22.0.txt]: Done https://github.com/pandas-dev/pandas/pull/18155/commits/21c5f5ea3528e451ec2653e19de094d25cdc0405

File: pandas/io/parquet.py [hunk 6]
```diff
@@ -76,9 +76,9 @@ def write(self, df, path, compression='snappy',
                 table, path, compression=compression,
                 coerce_timestamps=coerce_timestamps, **kwargs)
 
-    def read(self, path):
+    def read(self, path, columns=None):
         path, _, _ = get_filepath_or_buffer(path)
-        return self.api.parquet.read_table(path).to_pandas()
+        return self.api.parquet.read_table(path, columns).to_pandas()
```
Comment (jreback) [inline pandas/io/parquet.py]: pass columns as a kwarg to read_table and to_pandas

File: pandas/io/parquet.py [hunk 7]
```diff
@@ -201,4 +204,4 @@ def read_parquet(path, engine='auto', **kwargs):
     """
 
     impl = get_engine(engine)
-    return impl.read(path)
+    return impl.read(path, columns)
```
Comment (jreback) [inline pandas/io/parquet.py]: same

File: pandas/tests/io/test_parquet.py [hunk 8]
```diff
@@ -282,6 +282,18 @@ def test_compression(self, engine, compression):
         df = pd.DataFrame({'A': [1, 2, 3]})
         self.check_round_trip(df, engine, compression=compression)
 
+    def test_read_columns(self, engine, fp):
+        # GH18154
+        df = pd.DataFrame({'string': list('abc'),
+                           'int': list(range(1, 4))})
+
+        with tm.ensure_clean() as path:
```
Comment (jreback) [inline pandas/tests/io/test_parquet.py]: you don’t need the fp argument here: engine cycles thru both engines
use check_round_trip; pass in the expected (and the columns kwarg)

**Review by jreback — CHANGES_REQUESTED** [review]
see comments

**Comment by jreback** [comment]
you have a linting issue

File: pandas/io/parquet.py [hunk 9]
```diff
@@ -76,9 +76,9 @@ def write(self, df, path, compression='snappy',
                 table, path, compression=compression,
                 coerce_timestamps=coerce_timestamps, **kwargs)
 
-    def read(self, path):
+    def read(self, path, columns=None):
         path, _, _ = get_filepath_or_buffer(path)
-        return self.api.parquet.read_table(path).to_pandas()
+        return self.api.parquet.read_table(path, columns=columns).to_pandas()
```
Comment (jreback) [inline pandas/io/parquet.py]: i’d like to pass thru kwargs as well; these won’t be specific names args just pass thru to the engine to validate
for both fp and pyarrow 
could just be a simple test with row_groups

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/io/parquet.py (hunk 9 above)
Comment (hoffmann) [reply pandas/io/parquet.py]: Ok, i think it is good to pass explicit options like columns which are supported by both backends and also pass the kwargs to be able to provide additional engine specific args.

Have to look at the test case.

Comment (jreback) [reply pandas/io/parquet.py]: ok that’s fine
really want row_group support :) (next PR!)
also if u want: https://github.com/pandas-dev/pandas/issues/17102

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
ping on green

**Comment by hoffmann** [comment]
@jreback green. 

If it's ok I'd like to do the change to accept **kwargs in the read() function in a different pull request because it will require to rewrite https://github.com/pandas-dev/pandas/blob/master/pandas/tests/io/test_parquet.py#L191 to be able to handle **kwargs for to_parquet and read_parquet at the same time.

**Comment by jreback** [comment]
> If it's ok I'd like to do the change to accept **kwargs in the read() function in a different pull request because it will require to rewrite https://github.com/pandas-dev/pandas/blob/master/pandas/tests/io/test_parquet.py#L191 to be able to handle **kwargs for to_parquet and read_parquet at the same time.

yep ok by me

File: doc/source/io.rst [hunk 10]
```diff
@@ -4538,6 +4538,16 @@ Read from a parquet file.
 
    result.dtypes
 
+Read only certain columns of a parquet file. 
+
+.. ipython:: python
```
Comment (jreback) [inline doc/source/io.rst]: in next PR, can you add a version added tag here (for 0.21.1)

**Comment by jreback** [comment]
thanks!

**Pull request merged**